File: JINSEOP/프로그래머스_WEEK9/쿼드압축 후 개수 세기.cpp
#include <string>
#include <vector>
using namespace std;

void compress(int x, int y, int arr_size, vector<vector<int>>& arr, vector<int>& answer) {
    bool is_zero = true, is_one = true;
    
    for (int i = x; i < x + arr_size; i++) {
        for (int j = y; j < y + arr_size; j++) {
            if (arr[i][j] == 0) is_one = false;
            if (arr[i][j] == 1) is_zero = false;
        }
    }

    if (is_zero == true) {
        answer[0]++;
        return;
    }
    
    if (is_one == true) {
        answer[1]++;
        return;
    }

    compress(x, y, arr_size/2, arr, answer);
    compress(x, y + arr_size / 2, arr_size/2, arr, answer);
    compress(x + arr_size / 2, y + arr_size/2, arr_size / 2, arr, answer);
    compress(x + arr_size /2 , y, arr_size / 2, arr, answer);
}

vector<int> solution(vector<vector<int>> arr) {
    vector<int> answer(2, 0);

    compress(0, 0, arr.size(), arr, answer);
    return answer;
}

// 반드시 &arr로 넘겨주자.
// 안그럼 시간초과 남!

File: DONGUK/프로그래머스_WEEK 9/쿼드압축 후 개수 세기.py
answer = [0,0]

def check(arr, start_i, start_j, cnt):
    val = arr[start_i][start_j] # 0 or 1
    flag = True

    for i in range(start_i, start_i + cnt):
        for j in range(start_j, start_j + cnt):
            if arr[i][j] != val:
                flag = False
    
    
    if flag:
        answer[val] += 1

    else:
        h_cnt = cnt // 2
        check(arr, start_i, start_j, h_cnt)
        check(arr, start_i, start_j + h_cnt, h_cnt)
        check(arr, start_i + h_cnt, start_j, h_cnt)
        check(arr, start_i + h_cnt, start_j + h_cnt, h_cnt)

def solution(arr):
    
    check(arr, 0, 0, len(arr))
    return answer